### ISSUE TITLE
Execute code when workflow is prematurely terminated

### ISSUE THREAD

**Issue by ssougnez** [danielgerlag/workflow-core#411]
Hi,

here is my use case... I created a workflow that starts by creating some data in a database. Then, basically, an user has to approve/reject a task. If the user approves the task, the workflow completes but if he rejects it, the data previously created gets deleted from the database.

This works OK, but the user also has the possibility to terminate the workflow (instead of approving/rejecting the task) for whatever reason. However, if he chooses to do so, the date from the database won't never get deleted.

It would be nice to be able to do something like:

```
public class MyWorkflow : IWorkflow
{
    public void Build(IWorkflowBuilder<MyData> builder)
    {    
        builder
            .StartWith<RandomStep>()
                .OnTermination<CleanUpStep>
           .Then<OtherRandomStep>()
               .OnTermination<OtherCleanUpStep>
           .Then<CompleteWorkflow>();
    }
}
```

Basically, the idea would be to add a "OnTermination" step that would be executed when the workflows gets terminates prematurely. Of course, in the above example, `CleanUpStep` gets executed if the workflow is terminated before `OtherRandomStep` is executed while `OtherCleanUpStep` gets executed if the workflow is terminated after `OtherRandomStep` is executed. Basically, only the last `OnTermination` would get executed if the workflow is terminated prematurely (it's not executed when the workflow completes normally).

What do you think?

**Comment by danielgerlag** [comment]
I like the idea... it would work something like the `defer` statement in Go?
https://tour.golang.org/flowcontrol/12

**Comment by ssougnez** [comment]
It could yeah, but with one small difference I think. If the workflow completes as expected, the deferred step should not execute. If I take my example, I don't want the dates to be deleted from the database if everything goes OK, so yeah, like the defer in go but only if the workflow gets abruptly terminated.

**Comment by ssougnez** [comment]
Just out of curiosity, do you consider implementing this in a short term period or not?

I'm asking that because I'll need this feature quite soon and if you don't plan on adding it in the upcoming weeks, I have to create an alternate implementation while waiting for yours :-D

**Comment by danielgerlag** [comment]
I'm really busy for the next 8 - 10 weeks, so I don't see myself doing any significant open source contributions until then.
